cooked-validators/src/Cooked/Attack/DoubleSat.hs

```diff
@@ -38,387 +34,111 @@ office. They both leave satisfied."
-    attacker
-    AllCombinations
-
--- | Parameters for a double satisfaction attack that adds 'SpendsScript'
--- constraints.
-dsAddSsc ::
-  Traversal' TxSkel a ->
-  (MockChainSt -> a -> [SpendsScriptConstraint]) ->
+-- This attack consists in adding some extra constraints to a transaction, and
+-- hoping that the additional checks triggered by the extra inputs are fooled by
+-- what's already present on the transaction. Any extra value contained in the
+-- added constraints is then paid to the attacker.
+doubleSatAttack ::
+  Is k A_Traversal =>
+  -- | Each focus of this optic is a potential reason to add some extra
+  -- constraints.
+  --
+  -- As an example, one could go through the 'PaysScript' constraints for
+  -- validators of type @t@ with the following traversal:
+  --
+  -- > paysScriptConstraintsT % paysScriptConstraintTypeP @t
+  Optic' k is TxSkel a ->
+  -- | Which constraints to add, for each of the foci. There might be different
+  -- options for each focus, that's why the return value is a list.
+  --
+  -- Coninuing the example, for each of the focused 'PaysScript' constraints,
+  -- you might want to try adding some 'SpendsScript' constraints to the
+  -- transaction. Since it might be interesting to try different redeemers on
+  -- these extra 'SpendsScript' constraints, you can just provide a list of all
+  -- the options you want to try adding for a given 'PaysScript' that's already
+  -- on the transaction.
+  (MockChainSt -> a -> [Constraints]) ->
+  -- | The wallet of the attacker, where any surplus is paid to.
+  --
+  -- In the example, the extra value in the added 'SpendsScript' constraints
+  -- will be paid to the attacker.
   Wallet ->
-  SplitStrategy ->
-  DoubleSatParams a
-dsAddSsc optic extraSsc attacker splitStrategy =
-  DoubleSatParams
-    { dsOptic = optic,
-      dsExtraConstraints = \mcst a ->
+  -- | Since there are potentially many options for each of the foci, the
+  -- question is whether (and how) to combine additions that were triggered by
+  -- different foci.
+  --
+  -- In the example: Let's say that the unmodified transaction had three focused
+  -- 'PaysScript' constraints, and that you want to try 2, 3, and 5 options for
+  -- additional 'SpendsScript' constraints for each of them, respectively.
+  --
+  -- - If you want to try each additional 'SpendsScript' on its own modified
+  --   transaction, use 'AllSeparate'. Thus, there'll be 2+3+5=10 modified
+  --   transactions.
+  --
+  -- - If you want to try combining all options from one focus with all options
+  --   from all of the other foci, use 'TryCombinations'. Then, there'll be
+  --   (2+1)*(3+1)*(5+1)=72 possible combinations, if you include all of the
+  --   combinations where /at most/ three (one for each focus) extra constraints
+  --   are added. One of these combinations is of course the one where nothing
+  --   is added, and that one is omitted, which brings the grand total down to
+  --   71 options.
```

Comment:
I can't figure out from this paragraph what combination of constraints are applied in each modified transaction.

Reply:
In a similar vein, making sure I'm following... you get 72 because you try 0, 1, or 2 additional `SpendsScript` constraints for the first `PaysScript` constraint, then 0, 1, 2, or 3 for the second, etc.? Hence (2+1)*(3+1)*(5+1)